### PR TITLE
[5.4] Allow specific date formats for toArray and toJson

### PR DESCRIPTION
**Pull request by timacdonald**
**tldr;** This PR introduces the ability to manipulate how dates are cast to string when calling `toArray` or `toJson`. There is currently no way to manipulate the way these are output.

## Example

Take the following class:

```php
<?php

namespace App;

use Illuminate\Database\Eloquent\Model;

class Event extends Model
{
    protected $attributes = [
        'date' => '2020-01-12',
        'another_date' => '2020-01-12',
    ];

    protected $dates = [
        'date:Y-m-d',
        'another_date',
    ];
}
```

And the result:

screen shot 2017-08-03 at 1 08 26 pm

So the idea here is that you can specify the format that dates are in. This is related to, but different to  [previous PR](https://github.com/laravel/framework/pull/20295).

Now, this probably has wider possibilities than what I have implemented here. This was just a proof of concept in a [discussion thread on internals](https://github.com/laravel/internals/issues/669#issuecomment-319603476) about the idea of this. I threw it together as an example....and then just decided to PR it. It may need further work to be in a decent state, and perhaps could be implemented better - advice welcomed!

## Alternatives

It was also suggested that an array based format was adopted. I personally prefer sticking with the `date:format` style as it matches other things in the Laravel code base - like validation rules.

```php
protected $dates = [
    'date' => 'Y-m-d',
    'another_date',
];
```

Or there could just be a `dateFormats` property, but that is to close to `dateFormat` for my liking.

```php
protected $dateFormats = [
    'date' => 'Y-m-d',
];
```

OR OR chances are you'd want to reuse the format maybe once or twice, so you could set it up....

```php
protected $dateFormats = [
    'Y-m-d' => ['date', 'this_other_date_in_this_format'],
];
```

## Performance Downfalls

- Whenever `getDates()` is called, now the values are mapped over and have a string check done on them - even if your not utilising the date formats.

- For every date when casting `toArray` or `toJson` means a loop over all the date attributes.

## More Possibilities

This could have further implications to allow devs to specify the incoming format of dates. Discussion around this has already been had in the [PR I mentioned previously](https://github.com/laravel/framework/pull/20295).

i.e. We could alter the `setAttribute` method to do this under the hood

```php
protected function setDateAttribute($value)
{
    $this->attributes['date'] = Carbon::createFromFormat(
        $this->getDateFormat('date'), $value
    );
}
```

I'd be happy to PR this functionality in another PR, or if you wanted, add it to this PR.

## Tests

No tests added yet, but happy to if there is interest in the PR. Not a breaking change in its current state.

## Related Links
https://github.com/laravel/internals/issues/669
https://github.com/laravel/internals/issues/630
https://github.com/laravel/internals/issues/699
https://github.com/laravel/framework/pull/20295
https://stackoverflow.com/questions/44490512/eloquent-model-multiple-date-formats

### CODE REVIEW

**Comment by sisve** [comment]
I don't understand. Is this so that your database models can be responsible for the api result generation? Wouldn't that be against several philosophies about single-responsibility and do-not-do-crazy-things? Isn't it more maintainable to add a separate api-result-generation layer, something that [Dingo calls transformers](https://github.com/dingo/api/wiki/Transformers)?

How would your solution support two versions of the api requiring two different result schemas?

**Comment by timacdonald** [comment]
I do agree with you, and I personally use API transformers for API stuff most of the time - also, I always-do-crazy-things™ 🤠

Discussion:

All attributes are already passed through their specific mutators when you call `toJson`, thus the model is already responsible (to some degree) for the output of itself - via the `toJson` method.

With that in mind, I figure it makes sense that dates should also have this ability - which they do - but it's a weird grey area, because although you can, and generally would cast your dates to Carbon instances to work with in your codebase, there is no way to specify how those instances are cast to string during the `tojson` call - which is the point of casting to json - to get a string representation of the data.

I believe that Carbon should be able to set an instance by instance `$toStringFormat`, but unfortunately that is a static class method. But if is was not a class method, something similar would need to be implemented to ensure that each instance gets the appropriate format added. That might be by a custom mutator, or some built in magic like is suggested in this PR.

I don't think it's outside the realm of possibility that someone would create an API end-point that just returns the model as a json response. I know I've personally done this when I've needed to poll to see if a model had been updated by a background job. Was it best practice, no. Did it do the job, and is still doing the job - yes. Would I do it again, oh hell yea.

This also opens up how dates can be passed to the model. So a model would be able to interpret different date formats as needed. Although this is a separate topic, it does tie into the - should the model know how to interpret incoming data, and translate outgoing data.

Just my thoughts :) Sorry for the epic response and, I assume, all the typos...

**Comment by timacdonald** [comment]
Sorry, I totally didn't address your other question.

This is not intended to be a full API transformer approach to support multiple API schemas etc.

As, I'm sure, `toJson` isn't either. But if the point of to json is to get a string representation of the models data, which you can adjust via mutators, I think it makes sense to be able to reach in, fiddle with a carbon date output format, and pass it through during the process.

You could do a `getBirthdayToJson($value)` or something similar - type mutator just for these scenarios - but I'm not sure about that really.

**Comment by taylorotwell** [comment]
Already possible by overriding `serializeDate` on your model:

```php
    /**
     * Prepare a date for array / JSON serialization.
     *
     * @param  \DateTimeInterface  $date
     * @return string
     */
    protected function serializeDate(DateTimeInterface $date)
    {
        return $date->format('U');
    }
```

**Comment by timacdonald** [comment]
Thanks Taylor, unfortunately that only allows for a global format for all dates attributes, the concept here (as poorly implemented as it may be 😁) was to allow attribute-by-attribute formats.

Thanks for taking the time to review it and the feedback as well @sisve.

**Comment by torkiljohnsen** [comment]
Great effort @timacdonald 👍 

It would be great @taylorotwell if the date format could be configured instead of having to override the method. Nowadays, API consumers expect ISO 8601 format dates. It is tedious to override `serializeDate` on each and every model in an app.

Having `dateFormat` property on models define both the database and the return format is not very useful, when MySQL for instance supports only three types of date/time column types, while output of date/time has so many more facets.

Alternatively, if Laravel would respect the Carbon setting for dateTime format, or keep some global default for datetime formatting, that would be cool too, so you could just define this once, for instance:

```php
Carbon::setToStringFormat(DateTime::ATOM);
```

**Comment by krychu90** [comment]
> 
> 
> Already possible by overriding `serializeDate` on your model:
> 
> ```
>     /**
>      * Prepare a date for array / JSON serialization.
>      *
>      * @param  \DateTimeInterface  $date
>      * @return string
>      */
>     protected function serializeDate(DateTimeInterface $date)
>     {
>         return $date->format('U');
>     }
> ```

This could be good enough if _serializeDate_ method had second parameter - _$key_, like this:

```
    /**
     * Prepare a date for array / JSON serialization.
     *
     * @param  \DateTimeInterface  $date
     * @return string
     */
    protected function serializeDate(DateTimeInterface $date, $key)
    {
        if (key === 'my_date') return $date->format('U');
        return $date->format($this->getDateFormat());
    }
```
